sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c

```diff
@@ -0,0 +1,225 @@
+
+OTTF_DEFINE_TEST_CONFIG();
+
+/**
+ * OTP HW partition relative IFETCH offset in bytes.
+ *
+ * x = OTP_CTRL_PARAM_EN_SRAM_IFETCH_OFFSET (1728)
+ * y = OTP_CTRL_PARAM_HW_CFG_OFFSET (1664)
+ * IFETCH_OFFSET = (x - y) = 64
+ */
+static const uint32_t kOtpIfetchHwRelativeOffset =
+    OTP_CTRL_PARAM_EN_SRAM_IFETCH_OFFSET - OTP_CTRL_PARAM_HW_CFG_OFFSET;
+
+/**
+ * OTP can only be accessed by 32b aligned addresses. As `csrng_sw_app_read` is
+ * not aligned, we read from the previous aligned address and use the following
+ * offset in bits to access the `csrng_sw_app_read` byte.
+ */
+static const uint32_t kOtpCsrngFwReadBitOffset =
+    (OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET -
+     OTP_CTRL_PARAM_EN_SRAM_IFETCH_OFFSET) *
+    8;
+
+enum {
+  kExpectedOutputLen = 16,
+};
+
+/**
+ * Checks the CSRNG internal state against `expected` values.
+ *
+ * @param csrng A CSRNG handle.
+ * @param expected Expected CSRNG internal state.
+ */
+
+static void check_internal_state(const dif_csrng_t *csrng,
+                                 const dif_csrng_internal_state_t *expected) {
+  csrng_testutils_cmd_ready_wait(csrng);
+  dif_csrng_internal_state_t got;
+  CHECK_DIF_OK(
+      dif_csrng_get_internal_state(csrng, kCsrngInternalStateIdSw, &got));
+
+  CHECK(got.instantiated == expected->instantiated, "exp: %x, got: %x",
+        expected->instantiated, got.instantiated);
+  CHECK(got.reseed_counter == expected->reseed_counter, "exp: %x, got: %x",
+        expected->reseed_counter, got.reseed_counter);
+  CHECK(got.fips_compliance == expected->fips_compliance, "exp: %x, got: %x",
+        expected->fips_compliance, got.fips_compliance);
+
+  CHECK_ARRAYS_EQ(got.v, expected->v, ARRAYSIZE(expected->v),
+                  "CSRNG internal V buffer mismatch.");
+
+  CHECK_ARRAYS_EQ(got.key, expected->key, ARRAYSIZE(expected->key),
+                  "CSRNG internal K buffer mismatch.");
+}
+
+/**
+ * CTR DRBG Known-Answer-Test (KAT) for INSTANTIATE command.
+ */
+static void fips_instantiate_kat(const dif_csrng_t *csrng, bool fail_expected) {
+  LOG_INFO("%s", __func__);
+
+  CHECK_DIF_OK(dif_csrng_uninstantiate(csrng));
+  const dif_csrng_seed_material_t kEntropyInput = {
+      .seed_material = {0x73bec010, 0x9262474c, 0x16a30f76, 0x531b51de,
+                        0x2ee494e5, 0xdfec9db3, 0xcb7a879d, 0x5600419c,
+                        0xca79b0b0, 0xdda33b5c, 0xa468649e, 0xdf5d73fa},
+      .seed_material_len = 12,
+  };
+  csrng_testutils_cmd_ready_wait(csrng);
+
+  CHECK_DIF_OK(dif_csrng_instantiate(csrng, kDifCsrngEntropySrcToggleDisable,
+                                     &kEntropyInput));
+
+  const dif_csrng_internal_state_t kExpectedState = {
+      .reseed_counter = 1,
+      .v = {0x06b8f59e, 0x43c0b2c2, 0x21052502, 0x217b5214},
+      .key = {0x941709fd, 0xd8a25860, 0x861aecf3, 0x98a701a1, 0x0eb2c33b,
+              0x74c08fad, 0x632d5227, 0x8c52f901},
+      .instantiated = true,
+      .fips_compliance = false,
+  };
```

Comment:
How did you generate the input and expected state? Can you put a reference in a comment?

Reply:
This is a copy of another test, which I would be moving to a test utils as recomented by @moidx.
@moidx Would you have the references for this values?

Reply:
Hi @engdoreis, you can probably copy this docstring to the new location. It documents the data source for the test vectors. You can also tag a `// TODO(#6643):` for a follow up cleanup, and to address @marnovandermaas's comment.

cross-referencing: #6643.

https://github.com/lowRISC/opentitan/blob/42009f4a57e0e60ae808311485d15dbdc04aa1e6/sw/device/tests/csrng_kat_test.c#L106-L136

Reply:
Thanks @moidx